### PR TITLE
Create 5% test for refreshing ads that have been viewable for 30+ seconds

### PR DESCRIPTION
**Pull request by JonNorman**
## What does this change?
1. Updates the preexisting `refresh-on-resize` to refresh ad slots via our prebid entry point, rather than the DFP entry point. This means that we get up-to-date prebid header bidding information on the refreshed ad.

2. Creates a 5% server-side test wherein the variant cohort will request a refresh of the ad 30 seconds after the ad has become viewable. We use an [event](https://developers.google.com/doubleclick-gpt/reference#googletageventsimpressionviewableevent) tied to DFP's [Active View](https://support.google.com/dfp_premium/answer/4524488?hl=en&visit_id=1-636535364226239161-2940600065&rd=1) criteria:

> - For ads, 50% of the ad’s pixels are visible in the browser window for a continuous 1 second.
> - For larger ads (those greater than 242,000 pixels), 30% of the ad’s pixels are visible in the browser window.
> - For in-stream video ads, 50% of the ad’s pixels must be visible in the browser window for a continuous 2 seconds.

3. In either case (refreshing due to viewport changes or viewability thresholds) the additional targeting information `refreshed=true` will be passed up into the ad call.

**Considerations (for discussion here)**
1. Do we want to restrict the ad slots that can be refreshed? Should we refresh high-merchandising?
2. Do we want to check that the tab is active before refreshing an ad? Is there anything else that may want to stay our ad-refreshing fingers?

## What is the value of this and can you measure success?
Phat thtackth.

### CODE REVIEW

**Comment by ScottPainterGNM** [comment]
1) I don't think we should refresh merchandising personally as not sure of the benefit?
2) I believe we do, viewability only counts when the tab is active so my understanding is we can only refresh after 30 seconds when it meets that criteria. i.e. active tab and in view

**Comment by JonNorman** [comment]
@ScottPainterGNM I've amended the behaviour such that ads are only refreshed when the [document is active](https://developer.mozilla.org/en-US/docs/Web/API/Page_Visibility_API) - i.e. once the threshold of 30 seconds has been met, we refresh immediately or when the page is next viewed.

On the second point, I've come to think that we should refresh all ads, at least as a starter for 10. Adding more rules about which ads refresh and which don't is going to add more cognitive load and make things harder to reason about. If good slot-specific reasons can be given in the future then we can definitely add a filter.

This work invalidates the assumption of "one ad per slot per pageview" and so no doubt any subsequent analysis in this area will need to be mindful of this.

**Comment by ScottPainterGNM** [comment]
I'm happy with that reasoning, in terms of direct/merch etc. we can rely on DFP to make the appropriate decisions. Then tweak if an issue.

We are adding a value when refresh=true correct? That way from a reporting perspective we can discount these impressions for direct campaigns if the same thing loads more than once.

I would like to see what a fabric takeover does visually perhaps but happy with what you've said.

**Comment by JonNorman** [comment]
Let's check a fabric takeover and see how it looks!

>We are adding a value when refresh=true correct?

Yep, [here](https://github.com/guardian/frontend/pull/19087/files#diff-08a901a89aee4c3674d3f5bbb929cb8fR19)

**Comment by JonNorman** [comment]
I've updated it so that we lazy-load all the refreshed ads, rather than immediately loading them after 30 seconds.

File: static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js [hunk 1]
```diff
@@ -0,0 +1,34 @@
+// @flow
+import type { ImpressionViewableEvent } from 'commercial/types';
+
+import { Advert } from 'commercial/modules/dfp/Advert';
+import { getAdvertById } from 'commercial/modules/dfp/get-advert-by-id';
+import { refreshAdvert } from 'commercial/modules/dfp/load-advert';
+
+const viweabilityThresholdMs = 30000;
```
Comment (LATaylor-guardian) [inline static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js]: Small typo: s/viweabilityThresholdMs/viewabilityThresholdMs

File: static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+
+export const onSlotViewable = (event: ImpressionViewableEvent): void => {
+    const advert: ?Advert = getAdvertById(event.slot.getSlotElementId());
+    if (advert) {
+        const onDocumentVisible = () => {
+            if (!document.hidden) {
+                document.removeEventListener(
+                    'visibilitychange',
+                    onDocumentVisible
```
Comment (LATaylor-guardian) [inline static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js]: Is this correct? It's referring to itself.

File: static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+        setTimeout(() => {
+            if (document.hidden) {
+                document.addEventListener(
+                    'visibilitychange',
+                    onDocumentVisible
+                );
+            } else {
+                refreshAdvert(advert);
+            }
+        }, viweabilityThresholdMs);
```
Comment (LATaylor-guardian) [inline static/src/javascripts/projects/commercial/modules/dfp/on-slot-viewable.js]: Small typo again: s/viweabilityThresholdMs/viewabilityThresholdMs

**Comment by LATaylor-guardian** [comment]
Looks good to me 👍

**Review by rich-nguyen — APPROVED** [review]
Excellent!

**Pull request merged**